README
Xgremlin is a hardware and operating system independent version of the
data analysis program Gremlin for Fourier transform spectrometry.
Xgremlin runs on PCs and workstations that use the X11 window
system, including cygwin in Windows.  In Xgremlin few things have 
changed from Gremlin and someone familiar with Gremlin should 
immediately be able to work with Xgremlin.

Gremlin has two main input modes, an edit mode where commands
are entered at a command prompt and a graphical plot mode with
a different set of commands.  Xgremlin emulates these two input modes
by splitting the application main window into an upper pane which is
used for plotting and a lower pane used for command editing and
program output. Both windows remain visible all the time. After
startup Xgremlin is in edit mode and commands may be typed in at the
command prompt. To change over to plot mode either the button labeled
Plotmode or the Tab key must be pressed. Pressing either
of these again returns Xgremlin to 'edit' mode. The Xgremlin commands
in plot and edit mode are identical to those of Gremlin ( with a few
exceptions - the use of the Tab key being one of them ).

Xgremlin startup
----------------
One area where Xgremlin differs from Gremlin significantly is
initialization.  When Xgremlin is started it first reads a system wide
initialization file xgremlin.conf. The system wide initialization
file is used to establish defaults that are applicable to any user on
a system such as the printer that is to be used for screen
dumps. Xgremlin then processes the personal initialization file
.xgremlinrc in the home directory.  The personal initialization file
can be used to taylor Xgremlin to the liking of each individual user
e.g. to associate certain commands with buttons or to modify the colors
used in the plotting window.

GUI elements 
------------
In Xgremlin, frequently used commands can be accessed conveniently
through simple mouse clicks on buttons. All important plot mode
commands for moving around in a plot are accessible through the row of
buttons on top of the plotting window. Whereas the functions of the
top row of buttons are fixed, Xgremlin has also a row of buttons
between the plot and the edit window which may be programmed by the
user with the button command. This provides a convenient
shortcut for frequently used plot or edit commands. A online help
system can be called up by selecting the menu entry Help in the
Gremlins menu. The menu entry Phase/Tranform pops up an
interface to phase correction and Fourier transforms. This is
explained in greater detail in the Xgremlin tutorial.


Command line editing
--------------------
Xgremlin has a simple command line editor for the text window. 
The editor commands are reminiscent of those used in the much loved and
hated Unix editor Emacs:

 Left arrow key
 Move cursor to the left
 Right arrow key
 Move cursor to the right
 Page Up key
 Scroll text in edit window back by one page
 Page down key
 Scroll text in edit window forward by one page
 Up arrow key
 Recall previous command or commands
 Conrol key + 'a' key
 Move cursor to beginning of the command line (Anfang)
 Control key + 'e' key
 Move cursor to end of the command line (Ende)
 Control key + 'k' key
 Delete everything from cursor position to end of line
 Backspace key
 Remove character to left of cursor

Line lists
----------
Binary line list (.lin) files are supported by Xgremlin independent of
the computer type Xgremlin is running on. A line list file written on
a PC with an Intel processor is readable on a Sun workstation and vice 
versa. In more technical terms: the byte order of data in .lin files
is always identical to the byte order of data in .lin files produced
by the original Gremlin program (i.e. little endian byte order).

Documentation
-------------

All of the documentation is available in html format in the file
README.html.


ChangeLog
2015-06-11 Gill
       * version 4.7
      
       * Response command fixed  - declare rindex double precision
       * Area command checks 1st limit less than second
       * Adding C of G to linelist saves limits in the identification.

2015-03-03 Gill
       * version 4.6

       * Now using GitHub
       * Print out more digits on A constants for hfs write
       * Made sure hfs fitting routines don't wipe out holds
       * Fix hfs delete so that it doesn't delete lines that are not there.
       * Fix hfs disp so it only displays up to max. no. lines
       * hfs disp without params now displays whole hfs linelist.


2015-02-23 Gill
       * version 4.5

       * Phase correction fixed so it only picks up one point per line (set by swid variable).
       * Change command fixed so commands actually match documentation
       * Inline functions modified so they work with new gfortran
       * Documentation for phase correction and change commands updated
       * New tutorial added for spline fitting
       * Documentation updated for installation


2009-06-09 Gill
        
        * version 4.3

        * Center of gravity now holds all parameters
        * Active does not modify lines that are held
        * Hold now has option of negative number which releases hold
        * Documentation for Hold, Active, updated
        * Fitlines will no longer work on complex files
        * Removed tabs from strings - can cause failure
	* dump can now make file name that includes a '.'
        * Made default browser firefox rather than mosaic (!)
        * Assorted other bug fixes.


2009-04-09 Gill 
        * version 4.2
       
        * HFS routines modified for gfortran. Tested on OS X and Linux

2001-04-01  Gill  

	* version 4.1

	* incuded calculation of uncertainties of intensities.

2000-07-30  Ulfi  <ulf@puffin.woodland>

	* version 4.0

	* attempt to fix the 'writefile' command.

	* integrate Gill's (originally Bruce Pulliam's) 
	hyperfine structure fitting code. Still needs lots
	of polishing but works. All the work was done by Gill Nave.
	

2000-03-07  Ulfi  <ulf@puffin.woodland>

	* version 3.15

	* fixed the 'fits2nso' command to correctly translate
	FITS files from the McMath archive at Kitt Peak NSO.
	

2000-02-26  Ulfi  <ulf@puffin.woodland>

	* version 3.12 - 3.14

        * A 'getlines' command now also plots the line markers 
        in xgremlin's plotting area by default.

        * The auxiliary plot function 'pltaux' now clips
        horizontally which is needed for the overplotting
        of hyperfine structure patterns.

        * wavenumber correction (wavcorr) is applied in a 
        'writeasc' command.
        
	* add a 'set noglitch' (and 'set glitch') option to
	turn off the deglitching algorithm that is applied to
	an interferogram before a transform. Needed to make sure
	that the chirps from etalon fringes are not recogized
	as glitches in the data.

	* added a new 'index' set of commands for Fourier
	transform refractometry.
	

1999-10-11  Ulfi  <ulf@puffin.woodland>

	* version 3.10

	* Wider format for point of c.g. in (^)g commands to 
	avoid *****.

	* xgremlin can now be compiled with the Lahey/Fujitsu
	Fortran 95 compiler. Very nice compiler, actually and not
	expensive ($200.-). The new compiler uncovered four genuine
	bugs which I had so far overlooked. 
	

1999-10-05  Ulfi  <ulf@puffin.woodland>

	* version 3.9

	* Ooops. The 'standards' command ignored the tolerance when
	both tolerance and maximum deviation factor were specified
	on the command line. Fixed.
	

1999-10-02  Ulfi  <ulf@puffin.woodland>

	* version 3.8

	* The plot mode move command 'v' which moves a line now
	also copies the line id correctly.

	* Fixed a bug in the 'background' command. The background
	is now calculated correctly - previously, the wrong spline
	knots were used. (The effect was almost unnoticable, though.)
	

1999-09-24  Ulfi  <ulf@puffin.woodland>

	* version 3.7

	* Spruced up the 'standards' command. The command now creates a
	.cal file which contains all calibration results and a plot of
	deviations to make it easy to identify outliers and other
	anomalies.
	

1999-09-21  Ulfi  <ulf@puffin.woodland>

	* version 3.6

	* the 'response' command must now specify whether the intensities
	resulting from 'writelines' should be proportional to the photon
	flux or proportional to the intensity. Improved the documentation.

	* one more significant figure for the wavelengths in the
	'writelines' command.
	

1999-09-17  Ulfi  <ulf@puffin.woodland>

	* version 3.5

	* added a new plot mode command S+c (shift c) which finds and
	displays the center of an interferogram in r.

	* in phase plot mode, the low resolution spectrum is now
        normalized to 1 to make it easier to determine a good cutoff factor.
	

1999-09-11  Ulfi  <ulf@puffin.woodland>

	* version 3.4

	* reading headers with continued lines finally works.

	* the variables that control the spline phase fit can
	now be changed with the info command.

	* Chebychev polynomial phase fits up to 15-th order 
	(i.e. 16 terms) are now possible.
	

1999-09-09  Ulfi  <ulf@puffin.woodland>

	* version 3.3

	* finally managed to make spline phase correction fit
	interruptible with the 'Stop' button. 

	* nK,nM,nS are now read in correctly. They are needed 
	for backward compatibility with older versions of 
	gremlin which don't understand the 'fdr' variable.

	* fixed numerous bugs in the functions that write 
	header files.
	

1999-09-08  Ulfi  <ulf@puffin.woodland>

	* version 3.2

	* fixed the 'filter' command which broke when 'stretch'
	was fixed.

	* fixed the 'read' command to not omit the last point in
	a complex file. This was a long-standing bug in gremlin.
	

1999-09-07  Ulfi  <ulf@puffin.woodland>

	* version 3.1

	* Made spline fit function in phase correction interruptible
	with the stop button.

	* When switching to the spline fit mode, the spline fit
	is turned off by default.
	


Fsrc/read.f
c \gremlin\src\read.for   Ver. 2.2

c  Revision 2.29  93/06/22  Merged with updated Decomp version
c  Revision 2.28  92/05/22  stretched read added; use 'set stretch' first
c  Revision 2.27  91/11/09  'read wavelength' added; use modew as flag
c  Revision 2.26  91/08/23  After byte swapping, store again as integer
c  Revision 2.25  91/05/22  delw handling updated for nwpp=2
c  Revision 2.24  90/02/17  use nrtr+2, not 4  
c  Revision 2.23  90/01/19  Long integer read added; bocode=3 added
c  Revision 1.05  05/30/93  nwpp handling updated
c  Revision 1.04  05/16/93  Fix bug in npt truncation
c  Revision 1.02  05/07/93  Fix bug in altdata reads
c  Revision 1.01  03/27/93  Allow read of 2*nrtr
   
c read wstart [wstop] ["unit name"]
c read wstart npts ["unit name"]
c The following commands set flags, but do not actually read:
c read "offset"  noffset
c read "wavelength"  	(modew   = 1)
c read "wavenumber"  	(modew   =-1)
  
c     wstart(wstop) -  starting(stopping)  wavenumber(length)
c     npts -  the number of points to read, max of 2*nrtr (131072)
c     "unit name" - GRAMMY unit name
c     noffset is in bytes; brev causes byte reversal
c     "integer" allows the reading of files of binary integers
c     A read with no unit specified defaults to the "datain" unit
c     Valid names are: datain, dataout and altdata.
c     These commands read a chunk of data starting
c     at the wavenumber(length) you specify.
 
      subroutine read(r,tr)

      integer nbpw
      parameter (nbpw = 4)  ! 4 bytes per float, now fixed

      include 'datetc.h'
      include 'infmtn.h'
      include 'iounit.h'
      include 'linparms.h'
      include 'set.h'
      include 'temp.h'

      real r(*),tr(*)

      integer npts, i, modew
      double precision wstrt, wstp
      integer maxpts
      integer noffset, nerr

      noffset = 0
      modew = 0

* flag that r contains new data
      lfirst = .true.

c set failure flag, test syntax
      ipara = 1
      if(ifx .gt. 1 .or. ifl .gt. 2) go to 9995
c test for special read flags
      if (nstrng .eq. 1 .and. ifl .eq. 0) go to 1000
c find unit
      lu = datain
      if(nstrng .gt. 0) lu = ionum(alpha(1)(1:nalpha(1)),nalpha(1))
      if (lu .eq. -1) lu = datain
c 			error if not proper unit or unit not open
      if (.not.(lu .eq. datain .or. lu .eq. dataot .or.
     &           lu .eq. altdat )) go to 9997
      if (uopen(lu) .eq. 0) go to 9998

c check for wavelength or wavenumber ranges
      if (modew .gt. 0) then
         if (ifl .eq. 1) then
            xparb = xnum(1)
            call wtos
            xnum(1) = xparb
         endif
         if (ifl .eq. 2) then
            xparb = xnum(2)
            call wtos
            xnum(2) = xparb
         endif
      endif

c recall original delw, nwpp; use fboffs (file byte offset) and 
c bocode (byte order code) if they appear in .hdr file
      delw = disp 
      nwpp = nwppt
      noffset = fboffs
c find number of bytes to read (default to 2k points)
      maxpts = nrtr + 2
      npts = maxpts                 ! read as much as possible by default
      if (ifx .eq. 1) npts = inum(1)
      if (ifl .eq. 2) npts = nwpp*(1 + (xnum(2)-xnum(1))/delw)
      if ( npts .lt. 1 ) then
         call wrtstr(
     &   ' Error :  invalid wavelength range or # of points')
         return
      end if
      if (npts .gt. maxpts) then
	 call wrtstr(
     &   ' Warning :  requested # of points exceeds capacity of r.')
         npts = nrtr + 2
         write (wrtbuf,111) npts
 111     format(' --- # of points reduced to ',i8)
         call wrtstr(wrtbuf)
      endif
 
c find start of read
      wstrt = xnum(1)
      if (rdref .eq. 1) wstrt = wstrt - 0.5*delw*npts/nwpp
      if (wstrt .lt. wstart) wstrt = wstart
      if (wstrt .ge. wstop) go to 9996
      nstart = nwpp*(wstrt - wstart)/delw
      if (nwpp.eq.2 .and. mod(nstart,2).ne.0) nstart = nstart + 1
      wstrt = wstart + delw*nstart/nwpp
      nbstart = nbpw*nstart + noffset
c adjust end point if necessary
      wstp = wstrt + delw*(npts-1)/nwpp
      if (wstp .gt. wstop) then
         npts = 0.5 + nwpp * (1.0 + (wstop - wstrt)/delw)
         wstp = wstrt + delw*(npts-1)/nwpp
      end if
      nbytes = nbpw*npts

* store the absolute point in the file where the buffer starts
      nrefp = nbstart / (nwpp * nbpw)

      call readata(lu, nbstart, r(1), nbytes,nerr)
      if (nerr .ne. 0) go to 9999
      wref = wstrt
      nop = npts
      
* if requested, change the byte order.  
      if ( bocode .ne. mbocode ) then
         call chibord( r, nop )
      end if

c check length for fft routines
      call strtrn
      if (warn .and. (2**me) .ne. nop)
     &	   call wrtout(output,' Warning :  length is not a power of 2')
      call disper

c either scale (using rdfctr from set)
      if ( .not.rdnorm .and. rdscale.gt.0 ) then
         do i=1,nop
            r(i) = r(i)*rdfctr
         end do
         scal = 1.0
         if (verbos) then
            write (wrtbuf,9011) wref
            call wrtout(output,wrtbuf)
 9011       format(' Wref = ',f15.5)
         end if
      else
c or normalize 
         absmax = 0.0
         do i=1,nop
            tmp = abs(r(i))
            if (tmp .gt. absmax) absmax = tmp
         end do
         scal = absmax
         do i=1,nop
            r(i) = r(i)/scal
         end do
         if (absmax.lt.100.0 .and. verbos) then
            write(wrtbuf,9006) wref, absmax
            call wrtout(output,wrtbuf)
         end if
 9006    format(' Wref = ',f15.5,' Absmax = ',f15.7)
         if (absmax.ge.100.0 .and. verbos) then
            write(wrtbuf,9007) wref, absmax
            call wrtout(output,wrtbuf)
         end if
 9007    format(' Wref = ',f15.5,' Absmax = ',f15.0)
      end if

c     if "stretch" has been set, do it:
      if (istret .ge. 2) then
         inum(1) = istret
         ifx = 1
         call strtch(r,tr)
      endif

* zap old line list in buffer
      nol = 0
      nlgot = 0
      ipara = 0
      call linevis( 0 )         ! flag line markers invisible
      call llerase              ! wipe out old plotter line list

c	write (*,*) 'nwpp=',nwpp,' irvmode=',irvmode,' istret=',istret
c	write (*,*) 'nbpw=',nbpw,' noffset=',noffset,' delw=',delw
c	write (*,*) 'npts=',npts,' nbstart=',nbstart,' nbytes=',nbytes
c	write (*,*) ' rdnorm=',rdnorm

      return
      
c Set special read flags.  Noffset is in bytes
1000  continue	
      if (alpha(1)(1:6)  .eq. 'offset') noffset = inum(1)
      if (alpha(1)(1:10) .eq. 'wavelength') modew = 1
      if (alpha(1)(1:10) .eq. 'wavenumber') modew = -1
      return
        
c error returns
9995  call wrtstr(' Error :  READ syntax error')
      return
9996  call wrtstr(' Error :  requested data outside window--no read ')
      return
9997  call wrtstr(' Error :  unacceptable I/O Unit for READ')
      return
9998  call wrtstr(' Error :  Specified unit not open')
      return
9999  call wrtstr(' Error :  read failed')
      return
      end
